libyul/optimiser/ZeroByReturndatasizeReplacer.cpp

```diff
@@ -0,0 +1,164 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+
+#include <libyul/optimiser/ZeroByReturndatasizeReplacer.h>
+
+#include <libyul/backends/evm/EVMDialect.h>
+#include <libyul/optimiser/CallGraphGenerator.h>
+#include <libyul/optimiser/OptimiserStep.h>
+#include <libyul/optimiser/Semantics.h>
+#include <libyul/SideEffects.h>
+#include <libyul/Utilities.h>
+#include <libyul/AST.h>
+
+#include <libsolutil/Algorithms.h>
+#include <libsolutil/CommonData.h>
+
+#include <range/v3/view/enumerate.hpp>
+#include <range/v3/view/reverse.hpp>
+
+using namespace std;
+using namespace solidity;
+using namespace solidity::yul;
+
+namespace
+{
+bool functionContainsCall(Dialect const& _dialect, map<YulString, SideEffects> const& _functionSideEffects, FunctionCall const& _funCall) {
+	if (BuiltinFunction const* builtin = _dialect.builtin(_funCall.functionName.name))
+		if (!builtin->sideEffects.keepsReturndatasize)
+			return true;
```

Comment:
I find it rather weird that the side effect propagator does not seem to report the side-effects of builtins themselves...

Reply:
Sounds like we should change that?

Reply:
I'd need to see how it interacts with the other uses, but maybe, yeah... the only downside would be that the set is larger, but that's not a major issue I guess. And for example ``SideEffectsCollector::operator()(FunctionCall const& _functionCall)`` has to do the same as this helper here...

Reply:
Actually, I'll just use the ``SideEffectsCollector`` instead for this PR for now, then we can change this separately.

Reply:
Thinking about it, actually I'll revert this again :-D. The ``SideEffectsCollector`` will visit the function call arguments, which is not a problem, but not needed for this use case here, so it's just wasteful...